geomstats/geometry/sasaki_metric.py

```diff
@@ -0,0 +1,238 @@
+     * parallel vector fields along curves are orthogonal to their fibres, and
+     * its restriction to any tangent space is Euclidean.
+
+    Geodesic computations are realized via a discrete formulation of the geodesic
+    equation on TM that involve geodesics, parallel translation, and the curvature
+    tensor on the base manifold M (see [1] for details).
+
+    Parameters
+    ----------
+    metric : RiemannianMetric
+        Metric of the base manifold of the tangent bundle.
+    n_s : int
+        Number of steps used in time-discrete geodesic computations.
+        Optional. Default : 3.
+
+    References
+    ----------
+    .. [1] Muralidharan, P., & Fletcher, P. T. (2012, June).
+    Sasaki metrics for analysis of longitudinal data on manifolds.
+    In 2012 IEEE conference on computer vision and pattern recognition (pp. 1027-1034).
+    https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4270017/
+    """
+
+    def __init__(self, metric: RiemannianMetric, n_s=3):
+        self.metric = metric  # Riemannian metric of underlying space
+        self.n_s = n_s  # Number of discretization steps
+        shape = (2, gs.prod(metric.shape))
+
+        super(SasakiMetric, self).__init__(2 * metric.dim, shape=shape,
+                                           default_point_type='matrix')
+
+    def exp(self, tangent_vec, base_point, n_steps=None, **kwargs):
+        """Compute the Riemannian exponential of a point.
+
+        Exponential map at base_point of tangent_vec computed by
+        shooting a Sasaki geodesic using an Euler integration on TM.
+
+        Parameters
+        ----------
+        tangent_vec : array-like, shape=[..., 2, M.dim]
+            Tangent vector in TTM at the base point in TM.
+        base_point : array-like, shape=[..., 2, M.dim]
+            Point in the tangent bundle TM of manifold M.
+        n_steps : int
+            Number of discrete time steps.
+            Optional, default: self.n_s.
+
+        Returns
+        -------
+        exp : array-like, shape=[..., 2, M.dim]
+            Point on the tangent bundle TM.
+        """
+        # unflatten
+        bs_pts = gs.reshape(base_point, (-1, 2) + self.metric.shape)
+        tngs = gs.reshape(tangent_vec, bs_pts.shape)
+
+        metric = self.metric
+        par_trans = metric.parallel_transport
+        n_s = self.n_s if n_steps is None else n_steps
+        eps = 1 / n_s
+
+        v0, w0 = tngs[:, 0], tngs[:, 1]
+        p0, u0 = bs_pts[:, 0], bs_pts[:, 1]
+        p, u = p0, u0
+        for _ in range(n_s):
+            p = metric.exp(eps * v0, p0)
+            u = par_trans(u0 + eps * w0, p0, None, p)
+            v = par_trans(v0 - eps * (metric.curvature(u0, w0, v0, p0)), p0, None, p)
+            w = par_trans(w0, p0, None, p)
+            p0, u0 = p, u
+            v0, w0 = v, w
+
+        return gs.reshape(gs.array(list(zip(p, u))), base_point.shape)
+
+    def log(self, point, base_point, n_steps=None, **kwargs):
```

Comment:
I have some more comments on this, which I'll address after the changes in `geodesic_discrete`.